lib/chef/knife/bootstrap.rb

```diff
@@ -621,6 +621,9 @@ def connect!
           # FIXME: this should save the key to known_hosts but doesn't appear to be
           config[:ssh_verify_host_key] = :accept_new
           do_connect(connection_opts(reset: true))
+        elsif e.message == "Sorry, we are unable to detect your platform" # Train::PlatformDetectionFailed
+          ui.warn("Could not connect with 'root'. Please provide valid user_name")
```

Comment:
You should be checking the exception here, not the text.

Also though if platform detection fails even though train can login to the host, but train doesn't actually know how to detect the platform it is going to fail with an incorrect message about supplying a valid user name.

Also root looks hardcoded here.   Plus there could be a lot of other cases where SSH might fail.

Most likely this must be changed in train to support better error messages about why platform detection fails.

Reply:
you could do something like `ui.warn("Could not detect the platform (check your username #{config[:whatevertheusernameis]} is correct and pass the correct username with --whatvertheuserflagis)")`

and the other problem is that "root" is certainly wrong if the target is windows.

Reply:
Yes @lamont you are correct here. So we are working out on finding right solution and debugging more to give right fix for this issue. May today I will fix this up.

Reply:
The main issue was that its giving wrong error and it should not go for platform detection unless the connection is made.  

So I tried debugging few things like I ran following on irb
```irb(main):006:0> Net::SSH.start("104.211.52.56", "root", :user_known_hosts_file=>"/dev/null", :port=>22, :compression=>false, :compression_level=>0, :keepalive=>true, :keepalive_interval=>60, :timeout=>60, :auth_methods=>["none", "publickey"], :keys_only=>true, :keys=>["/home/msys/.ssh/id_rsa"], :password=>"<hidden>", :non_interactive=>true, :verify_host_key=>:accept_new) do |ssh|
irb(main):007:1*   result = ssh.exec!("ls -l")
irb(main):008:1>   puts result
irb(main):009:1> end
Please login as the user "azure" rather than the user "root".
```
Which is actually a right behavior and it gives error to use what is expected and same should be displayed to user but that's not happening so I am fixing this now.